### ISSUE TITLE
Detect excess spacing for groups of aligned tokens

### ISSUE THREAD

**Issue by backus** [rubocop-hq/rubocop#3945]
The `ExtraSpacing` cop is great but I think it could be improved. Basically, it does a great job at detecting extra spacing for a node that is not aligned with anything above or below. Sometimes though it is possible for a cluster of aligned tokens to all have unnecessary extra spacing and this isn't caught. Examples:

```ruby
# This is ideal
foo              { something      }
barbaz           { something_else }
long_method_name { one_more_thing }

# Entire group has excess leading spacing
foo                { something      }
barbaz             { something_else }
long_method_name   { one_more_thing }

# Entire group has excess spacing before the last brace
foo              { something        }
barbaz           { something_else   }
long_method_name { one_more_thing   }

# Entire group has excess spacing after the opening brace
foo              {    something      }
barbaz           {    something_else }
long_method_name {    one_more_thing }
```

This has been one of the biggest things I've wanted from rubocop for a while (flagging whitespace in code review sucks for the reviewer and the author) but for a while I figured that there might be some tricky reason why this isn't caught. I can't think of a reason why though so I figured I would take a stab at fixing this soon.

I'm not sure if this makes sense as an improvement to the `ExtraSpacing` cop or if it should be a separate new cop.

**Comment by bbatsov** [comment]
Yeah, that'd be cool. Sounds like a challenge for our Master of Whitespace, the almighty @jonas054! :-)

**Comment by backus** [comment]
Wow I tried to implement a spike of this for a few hours yesterday (almost got it but couldn't finish the implementation without breaking alignment of tokens separated by newlines and comments). Way harder than I expected! Might give it another try soon but I might have to leave it to @jonas054.

A few thoughts from trying to implement this:

* It seems like the `ForceEqualSignAlignment` option should just be a separate cop. The logic to handle equal sign alignment is mostly separate from the other code and it makes it harder to reason about the cop. Not saying anyone made a bad decision just pointing out a possible refactor that could make something like this easier.
* It seems like cop might benefit from doing some token + special node normalization before doing alignment. Special casing for things like unary `+` is kind of confusing
* Rewriting some of the `PrecedingFollowingAlignment` mixin to provide methods that take a token and return a group of tokens that are aligned with the input token would make this task much easier.

**Comment by jonas054** [comment]
@backus Have another go if it brings you pleasure. Let me know if you give up, and I'll take a look when I can fit it in. I find myself having less time for fun - and by that I mean working on RuboCop - than before.

**Comment by backus** [comment]
@jonas054 Consider me given up then. I have plenty of OSS things I want to return to eventually so realistically would take me a long time to get back to this issue

**Comment by owst** [comment]
I think this is a duplicate of https://github.com/bbatsov/rubocop/issues/3094 (where I also determined that this would be tricky to implement!)

**Comment by backus** [comment]
Heck, sorry about that. @owst even though that issue is older, what do you think about closing that one and adding your description here? This issue seems to have a bit more traction / context added and it has already been labeled and whatnot. Also I thi

**Comment by owst** [comment]
No worries, and @bbatsov has closed the older issue already. The only example from my issue that isn't explicitly mentioned here is:
```
foo(1, 2,  3)
foo(4, 2,  3)
```
where both `3`s could have 1 leading space removed.

**Comment by bbatsov** [comment]
I'm closing this ticket due to no recent activity. Feel free to re-open it if you ever come back to it.
